salt/cloud/clouds/nova.py

```diff
@@ -722,6 +722,159 @@ def request_instance(vm_=None, call=None):
+            return
+
+    access_ip = node.get('extra', {}).get('access_ip', '')
+
+    rcv3 = rackconnectv3(vm_) in node['addresses']
+    sshif = ssh_interface(vm_) in node['addresses']
+
+    if any((rcv3, sshif)):
+        networkname = rackconnectv3(vm_) if rcv3 else ssh_interface(vm_)
+        for network in node['addresses'].get(networkname, []):
+            if network['version'] is 4:
+                access_ip = network['addr']
+                break
+        vm_['cloudnetwork'] = True
+
+    # Conditions to pass this
+    #
+    #     Rackconnect v2: vm_['rackconnect'] = True
+    #         If this is True, then the server will not be accessible from the ipv4 addres in public_ips.
+    #         That interface gets turned off, and an ipv4 from the dedicated firewall is routed to the
+    #         server.  In this case we can use the private_ips for ssh_interface, or the access_ip.
+    #
+    #     Rackconnect v3: vm['rackconnectv3'] = <cloudnetwork>
+    #         If this is the case, salt will need to use the cloud network to login to the server.  There
+    #         is no ipv4 address automatically provisioned for these servers when they are booted.  SaltCloud
+    #         also cannot use the private_ips, because that traffic is dropped at the hypervisor.
+    #
+    #     CloudNetwork: vm['cloudnetwork'] = True
+    #         If this is True, then we should have an access_ip at this point set to the ip on the cloud
+    #         network.  If that network does not exist in the 'addresses' dictionary, then SaltCloud will
+    #         use the initial access_ip, and not overwrite anything.
+
+    if any((cloudnetwork(vm_), rackconnect(vm_))) and (ssh_interface(vm_) != 'private_ips' or rcv3) and access_ip != '':
+        data.public_ips = [access_ip, ]
+        return data
+
+    result = []
+
+    if 'private_ips' not in node and 'public_ips' not in node and \
+       'floating_ips' not in node and 'fixed_ips' not in node and \
+       'access_ip' in node.get('extra', {}):
+        result = [node['extra']['access_ip']]
+
+    private = node.get('private_ips', [])
+    public = node.get('public_ips', [])
+    fixed = node.get('fixed_ips', [])
+    floating = node.get('floating_ips', [])
+
+    if private and not public:
+        log.warning(
+            'Private IPs returned, but not public... Checking for '
+            'misidentified IPs'
+        )
+        for private_ip in private:
+            private_ip = preferred_ip(vm_, [private_ip])
+            if private_ip is False:
+                continue
+            if salt.utils.cloud.is_public_ip(private_ip):
+                log.warning('{0} is a public IP'.format(private_ip))
+                data.public_ips.append(private_ip)
+                log.warning(
+                    (
+                        'Public IP address was not ready when we last'
+                        ' checked.  Appending public IP address now.'
```

Comment:
@cachedout Actually extra parenthesis here is more disturbing, yet that wasn't my code. But I will fix that too. :)